File: src/passgen.rs
use rand::Rng;

#[derive(Debug, Clone)]
pub struct PasswordGenConfig {
    pub length: usize,
    pub use_uppercase: bool,
    pub use_lowercase: bool,
    pub use_numbers: bool,
    pub use_symbols: bool,
}

impl Default for PasswordGenConfig {
    fn default() -> Self {
        Self {
            length: 20,
            use_uppercase: true,
            use_lowercase: true,
            use_numbers: true,
            use_symbols: true,
        }
    }
}

impl PasswordGenConfig {
    pub const MIN_LENGTH: usize = 8;
    pub const MAX_LENGTH: usize = 100;

    pub fn set_length(&mut self, length: usize) {
        self.length = length.clamp(Self::MIN_LENGTH, Self::MAX_LENGTH);
    }

    fn get_charset(&self) -> Vec<char> {
        let mut charset = Vec::new();

        if self.use_lowercase {
            charset.extend("abcdefghijklmnopqrstuvwxyz".chars());
        }

        if self.use_uppercase {
            charset.extend("ABCDEFGHIJKLMNOPQRSTUVWXYZ".chars());
        }

        if self.use_numbers {
            charset.extend("0123456789".chars());
        }

        if self.use_symbols {
            charset.extend("!@#$%^&*()_+-=[]{}|;:,.<>?".chars());
        }

        charset
    }

    pub fn has_valid_options(&self) -> bool {
        self.use_uppercase || self.use_lowercase || self.use_numbers || self.use_symbols
    }
}

pub fn generate_password(config: &PasswordGenConfig) -> Result<String, String> {
    if !config.has_valid_options() {
        return Err("At least one character type must be selected".to_string());
    }

    let charset = config.get_charset();
    if charset.is_empty() {
        return Err("No character set available".to_string());
    }

    let mut rng = rand::rng();
    let password: String = (0..config.length)
        .map(|_| {
            let idx = rng.random_range(0..charset.len());
            charset[idx]
        })
        .collect();

    Ok(password)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_config() {
        let config = PasswordGenConfig::default();
        assert_eq!(config.length, 20);
        assert!(config.use_uppercase);
        assert!(config.use_lowercase);
        assert!(config.use_numbers);
        assert!(config.use_symbols);
    }

    #[test]
    fn test_generate_password_length() {
        let config = PasswordGenConfig::default();
        let password = generate_password(&config).unwrap();
        assert_eq!(password.len(), 20);
    }

    #[test]
    fn test_generate_password_min_length() {
        let mut config = PasswordGenConfig::default();
        config.set_length(5); // Should clamp to MIN_LENGTH
        assert_eq!(config.length, PasswordGenConfig::MIN_LENGTH);
    }

    #[test]
    fn test_generate_password_max_length() {
        let mut config = PasswordGenConfig::default();
        config.set_length(150); // Should clamp to MAX_LENGTH
        assert_eq!(config.length, PasswordGenConfig::MAX_LENGTH);
    }

    #[test]
    fn test_no_options_selected() {
        let config = PasswordGenConfig {
            length: 20,
            use_uppercase: false,
            use_lowercase: false,
            use_numbers: false,
            use_symbols: false,
        };
        let result = generate_password(&config);
        assert!(result.is_err());
    }
}

File: app/lib/passgen.dart
/// Password generator — pure-Dart port of `core/src/passgen.rs`.
///
/// Same character sets, same length clamp (8..100), same "at least one set"
/// rule. Uses `Random.secure()` for generation. UI lives in
/// `screens/password_generator_screen.dart`.
library;

import 'dart:math';

const String _lowercase = 'abcdefghijklmnopqrstuvwxyz';
const String _uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const String _numbers = '0123456789';
const String _symbols = r'!@#$%^&*()_+-=[]{}|;:,.<>?';

class PasswordGenConfig {
  PasswordGenConfig({
    this.length = 20,
    this.useUppercase = true,
    this.useLowercase = true,
    this.useNumbers = true,
    this.useSymbols = true,
  });

  static const int minLength = 8;
  static const int maxLength = 100;

  int length;
  bool useUppercase;
  bool useLowercase;
  bool useNumbers;
  bool useSymbols;

  void setLength(int value) =>
      length = value.clamp(minLength, maxLength);

  bool get hasValidOptions =>
      useUppercase || useLowercase || useNumbers || useSymbols;

  String get _charset {
    final buf = StringBuffer();
    if (useLowercase) buf.write(_lowercase);
    if (useUppercase) buf.write(_uppercase);
    if (useNumbers) buf.write(_numbers);
    if (useSymbols) buf.write(_symbols);
    return buf.toString();
  }
}

/// Thrown when no character class is enabled.
class PasswordGenException implements Exception {
  PasswordGenException(this.message);
  final String message;
  @override
  String toString() => 'PasswordGenException: $message';
}

String generatePassword(PasswordGenConfig config, {Random? rng}) {
  if (!config.hasValidOptions) {
    throw PasswordGenException('At least one character type must be selected');
  }
  final charset = config._charset;
  final r = rng ?? Random.secure();
  final buf = StringBuffer();
  for (var i = 0; i < config.length; i++) {
    buf.write(charset[r.nextInt(charset.length)]);
  }
  return buf.toString();
}
